xarray/plot/utils.py

```diff
@@ -304,9 +304,11 @@ def _determine_cmap_params(
     if extend is None:
         extend = _determine_extend(calc_data, vmin, vmax)
 
-    if levels is not None or isinstance(norm, mpl.colors.BoundaryNorm):
+    if levels is not None:
         cmap, newnorm = _build_discrete_cmap(cmap, levels, extend, filled)
         norm = newnorm if norm is None else norm
+    if isinstance(norm, mpl.colors.BoundaryNorm):
```

Comment:
I think there is some incompatibility between "levels" and "BoundaryNorm" - you really shouldn't use both and there should be no need to discretize the colormap if using a BoundaryNorm;  conversely, there is no need to use a BoundaryNorm _or_ discretize the colormap if using "levels" in a contour plot.  

I'm not fully familiar with the xarray pcolormesh API, but I would raise if the user passes both BoundaryNorm and "levels", as having incompatible or duplicated specifications of the color boundaries, and I wouldn't do anything if a BoundaryNorm is passed in.  If you want discrete colormaps with levels, but a non-linear norm, I think `_build_discrete_cmap` is fine.

Reply:
Thanks for your feedback. I see your point and I agree, since the "levels" can be derived from BoundaryNorm.vmin, BoundaryNorm.vmax and BoundaryNorm.N. Issues https://github.com/pydata/xarray/issues/4061 and https://github.com/Deltares/xugrid/issues/49 are fixed by supplying levels instead of norm to ds.plot() indeed.

However, if it is in that case not beneficial to supply norm=BoundaryNorm to ds.plot(), I would say it should raise an error instead of being in an if-statement. Would you agree?

Reply:
Assuming we are talking about `pcolormesh`, Matplotlib has no *levels* kwarg, so it's a bit strange for xarray to have grown one.  I would not prohibit the Matplotlib way of doing this (`BoundaryNorm`).   If you want to keep the convenience of *levels*, then I would just error if both *levels* and `BoundaryNorm` are passed.    

In either case, I'm not clear why folks are using levels at all for a pcolormesh, but that is just me ;-)

Reply:
@jklymak, the issue is that levels is constructed by xarray, it is not passed by the user. The user passes only a norm argument only and a cmap.

Reply:
If there is a BoundaryNorm already, I'm not following why you are wanting to build a discrete cmap - the norm will discretize the underlying cmap for you...

Reply:
@jklymak You can ignore the PR I did, this is a workaround for the issue in #4061, but as I have learned from you it is not an improvement for xarray. The issue I am referring to provides the norm argument upon plotting, but it behaves different than the matplotlib. I do not understand the xarray plotting code good enough to provide a proper fix for this issue since it is a bit complex. As far as I understand, this PR can be closed without merging, but the issue is not fixed yet. Maybe you can suggest a proper fix for the code in the issue? You seem to understand how the plotting methods should behave.

Reply:
I think this PR is very close, just you went the opposite way with it.  I think it should actually be 

```
if (levels is not None) and (not isinstance(norm, mpl.colors.BoundaryNorm)):
```

and not create a colormap if BoundaryNorm is being used....